### PR TITLE
Updating to 11.8.2

### PR DESCRIPTION
**Pull request by m3vaz**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by m3vaz** [comment]
@conda-forge-admin, please rerender

**Comment by jakirkham** [comment]
Thanks Mark! 🙏 

Think we can make a few changes to the recipe (based on Bradley's work). Have tried to describe them via diffs below

As `cudatoolkit` comes from `{{ compiler('cuda') }}`, we can just ignore `run_exports` from that package

```diff
-  ignore_run_exports:
-   - cudatoolkit
+   ignore_run_exports_from:
+    - {{ compiler('cuda') }}
```

Can generalize the `sysroot` requirement for all dependencies like so

```diff
 requirements:
   build:
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
     - {{ compiler('cuda') }}
-    - sysroot_linux-64 2.17  # [linux64]
+    - sysroot_{{ target_platform }} 2.17  # [linux]
```

Might be worth adopting `cuda-version` now (though we could also save this for the CUDA 12 PR)

```diff
   host:
     ...
+    - cuda-version {{ major_version }}.0
   run:
     ...
+    - {{ pin_compatible('cuda-version', max_pin='x') }}
```

As this comes from the `sysroot` requirement in `build`, we can drop the explicit `__glibc` requirement

```diff
-  run_constrained:
-    - __glibc >=2.17  # [linux]
```

**Comment by m3vaz** [comment]
@jakirkham ignore_run_exports and sysroot changes are fine, I'll push those into the PR in a few hours

IIRC cuda-version is a package to ensure cuda-* and cudatoolkit don't mix, right? Is it supposed to be added to a cudatoolkit based package? (Also should it be 11.0?)

**Comment by jakirkham** [comment]
Sounds good. Thanks Mark! 🙏

Yeah we backported the `cuda-version` package to all CUDA versions conda-forge has supported. Here's a [list of the versions built]( https://github.com/conda-forge/cuda-version-feedstock/blob/3ce1bae701079226f982b173667edc5d0a64c01f/recipe/conda_build_config.yaml#L1-L16 )

The `cuda-version` package [constrains `cudatoolkit`]( https://github.com/conda-forge/cuda-version-feedstock/blob/3ce1bae701079226f982b173667edc5d0a64c01f/recipe/meta.yaml#L17 ). So if say `cuda-version=11.0` were installed with `cudatoolkit`, then the `cudatoolkit=11.0` package would be installed. Same story for any other CUDA version that might be selected. With CUDA 12.0, this effectively introduces a conflict since there is not a `cudatoolkit=12.0` package (nor is one planned atm).

FWIW don't have strong feelings about whether to include the `cuda-version` changes. More noting it as an option. It may also minimize the differences between the `cuda-python` 11 & 12 versions (if more 11 versions need to be produced in the future)

**Comment by m3vaz** [comment]
@jakirkham Sounds reasonable, I've updated the recipe per your comments.

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:37]:
```suggestion
    - cuda-version {{ major_version }}
```

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:40]:
```suggestion
    - {{ pin_compatible('cuda-version', min_pin='x', max_pin='x') }}
```

**Review by jakirkham — COMMENTED** [review]
Thank Mark! 🙏

There were some [conda install conflicts on CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=716307&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=b2a8456a-fb11-5506-ca32-5ccd32538dc0&l=92 ) (also noted in the snippet below):

<details>

```
The reported errors are:
- Encountered problems while solving:
-   - package cuda-version-11.0-h6b8d8af_0 has constraint cudatoolkit 11.0|11.0.* conflicting with cudatoolkit-11.8.0-h37601d7_10
- 
```

</details>

Have noted a few minor tweaks to how `cuda-version` is specified below, which should address this issue

Otherwise this looks good

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:43]: Also think we can drop this. Sorry for not catching this in the review above

```suggestion
```

**Review by bdice — APPROVED** [review]
Applying of @jakirkham’s comments will resolve my questions here. Approving based on that.

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:40]: Think we still need `cudatoolkit` in the CUDA 11 case for `cudart` and `nvrtc`

Can now rely on `cuda-version` to constrain that dependency though

```suggestion
    - {{ pin_compatible('cuda-version', min_pin='x', max_pin='x') }}
    - cudatoolkit  # [major_version == "11"]
```

**Review by jakirkham — COMMENTED** [review]
Thanks all!

One last suggestion to keep `cudatoolkit` in the CUDA 11 case

**Comment by jakirkham** [comment]
Went ahead and committed that last tweak. Please let me know if that looks ok. Happy to discuss as needed

**Comment by m3vaz** [comment]
> Went ahead and committed that last tweak. Please let me know if that looks ok. Happy to discuss as needed

I didn't think I could use selectors with variables defined in the file itself. LGTM

**Review by jakirkham — COMMENTED** [review]
Unfortunately Jinja variables can't be used in selectors. Sorry for forgetting this

We can either do this or use a Jinja `if` block. No strong feelings between the two

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:41]: Here's the alternative using Jinja

```suggestion
    {% if major < 12 %}
    - cudatoolkit
    {% endif %}
```

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:5]: Suggesting adding Jinja filters to convert these to integers to enable comparisons

```suggestion
{% set major_version = version.split(".")[0]|int %}
{% set minor_version = version.split(".")[1]|int %}
{% set patch_version = version.split(".")[2]|int %}
```

**Review by jakirkham — COMMENTED** [review]
Here's the alternative with Jinja for comparison. Please let me know if you have a preference between the two. Happy with either personally

File: recipe/meta.yaml
Comment (bdice) [reply recipe/meta.yaml:41]: I'm happy with either approach as well. I think that if the package lists for CUDA 12 are more than a couple lines, I slightly prefer Jinja if-blocks over selectors on each line, just to avoid repeating the same selectors over and over.

File: recipe/meta.yaml
Comment (bdice) [reply recipe/meta.yaml]: Thanks. I needed this snippet yesterday. :)

File: recipe/meta.yaml
Comment (m3vaz) [reply recipe/meta.yaml:41]: +1 on the jinja if blocks

Comment (m3vaz) [reply recipe/meta.yaml:41]: Related question, since this is a run dependency, if someone has cuda-* 12.0 installed in their environment, what would happen if they tried to install cuda-python 11.8 (either directly or because a package depends on it)? Wouldn't this force the environment to resolve to cudatoolkit rather than cuda-*>=12 (which would also work)?

Comment (leofang) [reply recipe/meta.yaml:41]: `cuda-version` would cause a conflict here, so to install 11.8 packages to an already-12.0 env will force conda to resolve the conflict, I think in this case it'd force the env to downgrade to 11.8.

Comment (m3vaz) [reply recipe/meta.yaml:41]: (basically, should this be a run or a run_constrained?)

Comment (jakirkham) [reply recipe/meta.yaml:41]: Since `cudatoolkit` was in `run`, this should be too

Comment (jakirkham) [reply recipe/meta.yaml:41]: Anyways sounds like there is a preference for Jinja. Will go ahead and commit this. Happy to discuss anything else as needed

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:41]: Since early parsing of the recipe may not have defined the Jinja variable yet (as seen by the lint error), check the variable is defined first and then compare

```suggestion
    {% if major is defined and major < 12 %}
```

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: recipe/meta.yaml
Comment (m3vaz) [reply recipe/meta.yaml]: cudatoolkit was in run because we did not have cuda-* packages. Now the case is that *either* may be used to satisfy cuda-python. 

The ask here is how to actually reflect that or if there is no way to reflect that.

**Review by jakirkham — APPROVED** [review]

File: recipe/meta.yaml
Comment (bdice) [inline recipe/meta.yaml:43]: Per offline discussion, there is no way to satisfy `cuda-python` 11.x with a CUDA 12 installation since it looks for particular versions of nvrtc. So we can probably simplify this and rely on the `cuda-version` pinning to constrain to CUDA 11.

Separately, I don't see where `major` is defined (I see `major_version`) so this is probably a bug anyway.
```suggestion
    - cudatoolkit
```

Comment (jakirkham) [reply recipe/meta.yaml:43]: `major` is defined based on the `version`, which is `11` up top

In version 12 (with your PR), this branch would effectively be disabled

The idea here was to make it easier to potentially update both versions in `main` (should that be needed)

That said, don't have strong feeling about that

Comment (bdice) [reply recipe/meta.yaml:43]: I'm not sure if I see that being defined. I see
```
{% set major_version = version.split(".")[0]|int %}
```
but `major_version` is not the same as `major`.

Comment (bdice) [reply recipe/meta.yaml:43]: I think the ability to update either build in `main` is potentially good, but I don't think I can wrap my head around all the changes that could require once #33 is merged... would prefer to keep it simple for this PR.

Comment (jakirkham) [reply recipe/meta.yaml:43]: Sorry you are right. Think it is suppose to be

```suggestion
    {% if major_version < 12 %}
    - cudatoolkit
    {% endif %}
```

Though no strong feelings if we would rather drop

Comment (bdice) [reply recipe/meta.yaml:43]: My vote is to drop (we can apply my original suggestion above if so).

Comment (jakirkham) [reply recipe/meta.yaml:43]: Sounds reasonable. Maybe we can wait for CI to finish. Commit your initial suggestion. Then merge?

**Comment by jakirkham** [comment]
Thanks all! 🙏

**Pull request merged**
